### PR TITLE
Support days mentioned only as ordinal numbers in CalendarMode+EN+.NET

### PR DESCRIPTION
**Pull request by STayinloves**
For cases like "between the 21st and 23rd", "week of the 18th" and "the 7th"

- minor fix typo SetParser
- fix bugs in datetimealt extractor

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.DateTime/Chinese/Parsers/MergedDateTimeParserChs.cs [hunk 1]
```diff
@@ -77,7 +77,7 @@ public new ParseResult Parse(ExtractResult er, DateObject refTime)
             }
             else if (er.Type.Equals(Constants.SYS_DATETIME_SET))
             {
-                pr = this.Config.GetParser.Parse(er, referenceTime);
+                pr = this.Config.SetParser.Parse(er, referenceTime);
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Chinese/Parsers/MergedDateTimeParserChs.cs]: :)

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs [hunk 2]
```diff
@@ -28,24 +28,26 @@ public List<ExtractResult> Extract(string text, DateObject reference)
...
+            var ordinalExtractions = config.OrdinalExtractor.Extract(text);
+
             tokens.AddRange(MergeTwoTimePoints(text, reference));
             tokens.AddRange(MatchDuration(text, reference));
-            tokens.AddRange(SingleTimePointWithPatterns(text, reference));
+            tokens.AddRange(SingleTimePointWithPatterns(text, new List<ExtractResult>(ordinalExtractions), reference));
             tokens.AddRange(MatchComplexCases(text, simpleCasesResults, reference));
             tokens.AddRange(MatchYearPeriod(text, reference));
-            tokens.AddRange(MatchOrdinalNumberWithCenturySuffix(text, reference));
+            tokens.AddRange(MatchOrdinalNumberWithCenturySuffix(text, new List<ExtractResult>(ordinalExtractions)));
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs]: This one really doesn't need "reference"?

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs [hunk 3]
```diff
@@ -220,20 +220,24 @@ private List<Token> MergeMultipleExtractions(string text, List<ExtractResult> ex
...
+            // For cases like "week of the 18th"
+            datePoints.AddRange(ordinalExtractions.Where(o => !datePoints.Any(er => er.IsOverlap(o))));
+
+            if (datePoints.Count < 1)
             {
                 return ret;
             }
 
-            foreach (var extractionResult in er)
+            foreach (var extractionResult in datePoints)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs]: Thanks a lot for these renames. They really improve code readability.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 4]
```diff
@@ -1370,7 +1368,7 @@ private DateTimeResolutionResult ParseWeekOfYear(string text, DateObject referen
                 targetWeekMonday = lastDayWeekMonday;
                 weekNum = Cal.GetWeekOfYear(targetWeekMonday, CalendarWeekRule.FirstFourDayWeek, DayOfWeek.Monday);
 
-                ret.Timex = $"{year.ToString("D4")}-{targetWeekMonday.Month.ToString("D2")}-W{weekNum}";
+                ret.Timex = $"{year:D4}-{targetWeekMonday.Month:D2}-W{weekNum}";
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: I'd prefer if the generation of timex expressions goes to it's own method.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 5]
```diff
@@ -1387,7 +1385,7 @@ private DateTimeResolutionResult ParseWeekOfYear(string text, DateObject referen
                 targetWeekMonday = firstDayWeekMonday.AddDays(7 * (cardinal - 1));
                 var targetWeekSunday = targetWeekMonday.This(DayOfWeek.Sunday);
 
-                ret.Timex = $"{year.ToString("D4")}-{targetWeekSunday.Month.ToString("D2")}-W{cardinal.ToString("D2")}";
+                ret.Timex = $"{year:D4}-{targetWeekSunday.Month:D2}-W{cardinal:D2}";
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Same.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 6]
```diff
@@ -1875,11 +1882,6 @@ private DateTimeResolutionResult ParseDecade(string text, DateObject referenceDa
             return ret;
         }
 
-        public bool GetInclusiveEndPeriodFlag()
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Why did you remove this?

File: Specs/DateTime/English/DateTimeModel.json [hunk 7]
```diff
@@ -5900,5 +5900,106 @@
         }
       }
     ]
+  },
+  {
+    "Input": "Cortana, can you please set something up for the week of the 18th.",
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModel.json]: Can you add some "week of the" case to the DateRange specs?

File: Specs/DateTime/English/DateTimeModel.json [hunk 8]
```diff
@@ -5900,5 +5900,106 @@
...
+              "value": "2018-09-04"
+            }
+          ]
+        }
+      }
+    ]
+  },
+  {
+    "Input": "Cortana, can you please set something up between the 21st and 23rd.",
+    "Context": {
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModel.json]: Can you add a "Comment" attribute here to highlight such cases are currently only supported in CalendarMode?

File: Specs/DateTime/English/DateTimeModel.json [hunk 9]
```diff
@@ -5900,5 +5900,106 @@
...
+      "ReferenceDateTime": "2018-08-08T10:00:00"
+    },
+    "Results": []
+  },
+  {
+    "Input": "Cortana, can you please set something up the 21st.",
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModel.json]: Same.

File: Specs/DateTime/English/DateTimeModelExtendedTypes.json [hunk 10]
```diff
@@ -1507,5 +1507,77 @@
...
+              "value": "2018-09-04"
+            }
+          ]
+        }
+      }
+    ]
+  },
+  {
+    "Input": "Cortana, can you please set something up between the 21st and 23rd.",
+    "Context": {
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModelExtendedTypes.json]: Same.

File: Specs/DateTime/English/DateTimeModelExtendedTypes.json [hunk 11]
```diff
@@ -1507,5 +1507,77 @@
...
+      "ReferenceDateTime": "2018-08-08T10:00:00"
+    },
+    "NotSupported": "javascript, python",
+    "Results": []
+  },
+  {
+    "Input": "Cortana, can you please set something up the 21st.",
+    "Context": {
+      "ReferenceDateTime": "2018-08-08T10:00:00"
```
Comment (tellarin) [inline Specs/DateTime/English/DateTimeModelExtendedTypes.json]: Same.

File: .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs (hunk 2 above)
Comment (STayinloves) [reply .NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs]: It doesn't need the reference for now, we can add it later if we need.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs (hunk 6 above)
Comment (STayinloves) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: This code block is not be used.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs (hunk 4 above)
Comment (STayinloves) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: I agree, but there is a lot of code of this kind need to be refactored, we can arrange a sperate PR to fix it.

File: Specs/DateTime/English/DateTimeModel.json (hunk 7 above)
Comment (STayinloves) [reply Specs/DateTime/English/DateTimeModel.json]: Added.
